packages/electron-commons/src/node-entry.ts

```diff
@@ -1,78 +1,112 @@
-import { BaseHost, COM, Communication } from '@wixc3/engine-core';
+import { BaseHost, COM, Communication, reportError } from '@wixc3/engine-core';
 import {
-    MetadataCollectionAPI,
-    metadataApiToken,
     ENGINE_ROOT_ENVIRONMENT_ID,
     METADATA_PROVIDER_ENV_ID,
+    MetadataCollectionAPI,
+    metadataApiToken,
 } from '@wixc3/engine-core-node';
-import { importModules, runIPCEnvironment } from '@wixc3/engine-runtime-node';
+import { emitEvent, importModules, runIPCEnvironment } from '@wixc3/engine-runtime-node';
 
-import { isNodeEnvStartupMessage } from './types';
+import { toError } from '@wixc3/common';
+import { NodeEnvironmentCommand, NodeEnvironmentEvent, NodeEnvironmentStartupCommand } from './types';
 
-const onMessageListener = async (message: unknown) => {
-    if (isNodeEnvStartupMessage(message)) {
-        const {
-            requiredModules,
-            basePath,
-            environmentName,
-            config,
-            environmentContextName,
-            featureName,
-            features,
-            bundlePath,
-            runtimeOptions,
-            parentEnvName,
-            env,
-        } = message.runOptions;
-        if (requiredModules) {
-            await importModules(basePath, requiredModules);
-        }
+let disposeEnv: (() => Promise<void>) | undefined;
```

Comment:
switch to a Set that contains openEnvs
iterate over it on close
clear it once done